src/components/ec/cuda/kernel/ec_cuda_reduce.cu

```diff
@@ -0,0 +1,340 @@
+                                cudaStream_t                 stream)
+{
+    int                th = EC_CUDA_CONFIG->reduce_num_threads;
+    unsigned long      bk = EC_CUDA_CONFIG->reduce_num_blocks;
+    ucc_reduction_op_t op;
+    ucc_datatype_t     dt;
+    size_t             count;
+
+    if (task->task_type == UCC_EE_EXECUTOR_TASK_REDUCE) {
+        dt    = task->reduce.dt;
+        count = task->reduce.count;
+        op    = task->reduce.op;
+    } else {
+        ucc_assert(task->task_type == UCC_EE_EXECUTOR_TASK_REDUCE_STRIDED);
+        dt    = task->reduce_strided.dt;
+        count = task->reduce_strided.count;
+        op    = task->reduce_strided.op;
+    }
+
+    if (count == 0) {
+        return UCC_OK;
+    }
+
+    bk = ucc_min((count + th - 1) / th, bk);
+
+    switch (dt) {
+    case UCC_DT_INT8:
+        DT_REDUCE_INT(int8_t, task, op, stream, bk, th);
+        break;
+    case UCC_DT_INT16:
+        DT_REDUCE_INT(int16_t, task, op, stream, bk, th);
+        break;
+    case UCC_DT_INT32:
+        DT_REDUCE_INT(int32_t, task, op, stream, bk, th);
+        break;
+    case UCC_DT_INT64:
+        DT_REDUCE_INT(int64_t, task, op, stream, bk, th);
+        break;
+    case UCC_DT_UINT8:
+        DT_REDUCE_INT(uint8_t, task, op, stream, bk, th);
+        break;
+    case UCC_DT_UINT16:
+        DT_REDUCE_INT(uint16_t, task, op, stream, bk, th);
+        break;
+    case UCC_DT_UINT32:
+        DT_REDUCE_INT(uint32_t, task, op, stream, bk, th);
+        break;
+    case UCC_DT_UINT64:
+        DT_REDUCE_INT(uint64_t, task, op, stream, bk, th);
+        break;
+    case UCC_DT_FLOAT16:
+        DT_REDUCE_FLOAT(__half, task, op, stream, bk, th);
+        break;
+    case UCC_DT_FLOAT32:
+#if SIZEOF_FLOAT == 4
+        DT_REDUCE_FLOAT(float, task, op, stream, bk, th);
+        break;
+#else
+        return UCC_ERR_NOT_SUPPORTED;
+#endif
+    case UCC_DT_FLOAT64:
+#if SIZEOF_DOUBLE == 8
+        DT_REDUCE_FLOAT(double, task, op, stream, bk, th);
+        break;
+#else
+        return UCC_ERR_NOT_SUPPORTED;
+#endif
+    case UCC_DT_FLOAT32_COMPLEX:
+#if SIZEOF_CUFLOATCOMPLEX == 8
+        DT_REDUCE_FLOAT_COMPLEX(cuFloatComplex, float, task, op, stream, bk, th);
+        break;
+#else
+        return UCC_ERR_NOT_SUPPORTED;
+#endif
+    case UCC_DT_FLOAT64_COMPLEX:
+#if SIZEOF_CUDOUBLECOMPLEX == 16
+        DT_REDUCE_FLOAT_COMPLEX(cuDoubleComplex, double, task, op, stream, bk, th);
+        break;
+#else
+        return UCC_ERR_NOT_SUPPORTED;
+#endif
+#if CUDART_VERSION >= 11000
+    case UCC_DT_BFLOAT16:
+        ucc_assert(2 == sizeof(__nv_bfloat16));
+        DT_REDUCE_FLOAT(__nv_bfloat16, task, op, stream, bk, th);
+        break;
+#endif
+    default:
+        ec_error(&ucc_ec_cuda.super, "unsupported reduction type (%s)",
+                 ucc_datatype_str(dt));
+        return UCC_ERR_NOT_SUPPORTED;
+    }
+    CUDA_CHECK(cudaGetLastError());
+    CUDA_CHECK(cudaStreamSynchronize(stream));
```

Comment:
cudaStreamSynchronize is not needed, if executor is the only interface for reductions then ucc_executor_task_test should indicate about completion. Removing stream synchronize will make reduce async

Reply:
done